crates/ruff_linter/src/rules/pycodestyle/rules/blank_lines.rs

```diff
@@ -0,0 +1,911 @@
+use itertools::Itertools;
+use std::cmp::Ordering;
+use std::num::NonZeroU32;
+use std::slice::Iter;
+
+use ruff_diagnostics::AlwaysFixableViolation;
+use ruff_diagnostics::Diagnostic;
+use ruff_diagnostics::Edit;
+use ruff_diagnostics::Fix;
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_codegen::Stylist;
+use ruff_python_parser::lexer::LexResult;
+use ruff_python_parser::lexer::LexicalError;
+use ruff_python_parser::Tok;
+use ruff_python_parser::TokenKind;
+use ruff_source_file::{Locator, UniversalNewlines};
+use ruff_text_size::TextRange;
+use ruff_text_size::TextSize;
+
+use crate::checkers::logical_lines::expand_indent;
+use crate::line_width::IndentWidth;
+use ruff_python_trivia::PythonWhitespace;
+
+/// Number of blank lines around top level classes and functions.
+const BLANK_LINES_TOP_LEVEL: u32 = 2;
+/// Number of blank lines around methods and nested classes and functions.
+const BLANK_LINES_METHOD_LEVEL: u32 = 1;
+
+/// ## What it does
+/// Checks for missing blank lines between methods of a class.
+///
+/// ## Why is this bad?
+/// PEP 8 recommends the use of blank lines as follows:
+/// - Two blank lines are expected between functions and classes
+/// - One blank line is expected between methods of a class.
+///
+/// ## Example
+/// ```python
+/// class MyClass(object):
+///     def func1():
+///         pass
+///     def func2():
+///         pass
+/// ```
+///
+/// Use instead:
+/// ```python
+/// class MyClass(object):
+///     def func1():
+///         pass
+///
+///     def func2():
+///         pass
+/// ```
+///
+/// ## References
+/// - [PEP 8](https://peps.python.org/pep-0008/#blank-lines)
+/// - [Flake 8 rule](https://www.flake8rules.com/rules/E301.html)
+#[violation]
+pub struct BlankLineBetweenMethods {
+    actual_blank_lines: u32,
+}
+
+impl AlwaysFixableViolation for BlankLineBetweenMethods {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let BlankLineBetweenMethods {
+            actual_blank_lines: nb_blank_lines,
+        } = self;
+        format!("Expected {BLANK_LINES_METHOD_LEVEL:?} blank line, found {nb_blank_lines}")
+    }
+
+    fn fix_title(&self) -> String {
+        "Add missing blank line(s)".to_string()
```

Comment:
Isn't this always going to be a single blank line (for this rule)? Can't this suggest removing blank lines too?

Reply:
> Isn't this always going to be a single blank line (for this rule)? 

Yes, I've fixed this in d188e200e.

> Can't this suggest removing blank lines too?

This is handled by a different rule (`E303`).